sway-ir/src/irtype.rs

```diff
@@ -412,43 +431,88 @@ impl Type {
         }
     }
 
-    pub fn size_in_bytes(&self, context: &Context) -> u64 {
+    /// Get the offset, in bytes, and the [Type] of the struct field at the index `field_idx`, if `self` is a struct,
+    /// otherwise `None`.
+    /// Panics if the `field_idx` is out of bounds.
+    pub fn get_struct_field_offset_and_type(
+        &self,
+        context: &Context,
+        field_idx: u64,
+    ) -> Option<(u64, Type)> {
+        if !self.is_struct(context) {
+            return None;
+        }
+
+        let field_idx = field_idx as usize;
+        let field_types = self.get_field_types(context);
+        let field_offs_in_bytes = field_types
+            .iter()
+            .take(field_idx)
+            .map(|field_ty| {
+                // Struct fields are aligned to word boundary.
+                field_ty.size(context).in_bytes_aligned()
+            })
+            .sum::<u64>();
+
+        Some((field_offs_in_bytes, field_types[field_idx]))
+    }
+
+    /// Get the offset, in bytes, and the [Type] of the union field at the index `field_idx`, if `self` is a union,
+    /// otherwise `None`.
+    /// Panics if the `field_idx` is out of bounds.
+    pub fn get_union_field_offset_and_type(
+        &self,
+        context: &Context,
+        field_idx: u64,
+    ) -> Option<(u64, Type)> {
+        if !self.is_union(context) {
+            return None;
+        }
+
+        let field_idx = field_idx as usize;
+        let field_type = self.get_field_types(context)[field_idx];
+        let union_size_in_bytes = self.size(context).in_bytes();
+        let field_size_in_bytes = field_type.size(context).in_bytes();
+
+        // The union fields are at offset (union_size - field_size) due to left padding.
+        Some((union_size_in_bytes - field_size_in_bytes, field_type))
+    }
+
+    pub fn size(&self, context: &Context) -> TypeSize {
```

Comment:
I hate long function names and this of all things is probably the best thing to call "size" without qualification, but you should either have a doc comment that clarifies we're talking about the size in the memory representation or make the name more explicit. I think the former is best.